algoliasearch/types_rule.go

```diff
@@ -0,0 +1,93 @@
+package algoliasearch
+
+type Rule struct {
+	ObjectID    string          `json:"objectID,omitempty"`
+	Condition   RuleCondition   `json:"condition"`
+	Consequence RuleConsequence `json:"consequence"`
+	Description string          `json:"description,omitempty"`
+}
+
+// RuleCondition is the part of an Algolia Rule which describes the condition
+// for the rule. The `Context` and `Priority` are optional, hence, they will
+// get ignored if an empty string or a 0 values are respectively set to them.
+type RuleCondition struct {
+	Anchoring RulePatternAnchoring `json:"anchoring"`
+	Pattern   string               `json:"pattern"`
+	Context   string               `json:"context,omitempty"`
+	Priority  int                  `json:"priority,omitempty"`
+}
+
+type RulePatternAnchoring string
+
+const (
+	Is         RulePatternAnchoring = "is"
+	StarstWith RulePatternAnchoring = "startsWith"
+	EndsWith   RulePatternAnchoring = "endsWith"
+	Contains   RulePatternAnchoring = "contains"
+)
+
+// NewSimpleRuleCondition generates a RuleCondition where only the `Anchoring`
+// and `Pattern` fields are specified. The optional `Context` and `Priority`
+// are then excluded.
+func NewSimpleRuleCondition(anchoring RulePatternAnchoring, pattern string) RuleCondition {
+	return NewRuleCondition(anchoring, pattern, "", 0)
+}
+
+// NewRuleCondition generates a RuleCondition where all the possible fields can
+// be specified.
+func NewRuleCondition(anchoring RulePatternAnchoring, pattern, context string, priority int) RuleCondition {
+	return RuleCondition{
+		Anchoring: anchoring,
+		Pattern:   pattern,
+		Context:   context,
+		Priority:  priority,
+	}
+}
+
+type RuleConsequence struct {
+	Params   *RuleConsequenceParams `json:"params,omitempty"`
+	Promote  []PromotedObject       `json:"promote,omitempty"`
+	UserData interface{}            `json:"userData,omitempty"`
+}
+
+type RuleConsequenceParams struct {
+	Query                         interface{} `json:"query,omitempty"`
+	AutomaticFacetFilters         []string    `json:"automaticFacetFilters,omitempty"`
+	AutomaticOptionalFacetFilters []string    `json:"automaticOptionalFacetFilters,omitempty"`
+}
+
+type QueryIncrementalEdit struct {
+	Remove []string `json:"remove"`
+}
+
+type PromotedObject struct {
+	ObjectID string `json:"objectID"`
+	Position int    `json:"position"`
+}
+
+type SaveRuleRes struct {
+	TaskID    int    `json:"taskID"`
+	UpdatedAt string `json:"updatedAt"`
+}
+
+type BatchRulesRes struct {
+	TaskID    int    `json:"taskID"`
+	UpdatedAt string `json:"updatedAt"`
+}
+
+type DeleteRuleRes struct {
+	TaskID    int    `json:"taskID"`
+	UpdatedAt string `json:"updatedAt"`
+}
+
+type ClearRulesRes struct {
+	TaskID    int    `json:"taskID"`
+	UpdatedAt string `json:"updatedAt"`
+}
+
+type SearchForRulesRes struct {
```

Comment:
What happens if the API suddenly returns more fields in the response? Is it going to cause the existing code to crash/return an error? Or will the extraneous fields simply be ignored? I am concerned about forward compatibility (i.e. making deployed code work with new versions of the API).